### ISSUE TITLE
lsetxattr() failed: Operation Not Permitted

### ISSUE THREAD

**Issue by Batcastle** [calamares/calamares#1237]
When attempting to install, Calamares throws an rsync error like such:
```
rync: rsync_xal_set: lsetxattr(...,"security.selinux") failed: Operation Not Permitted
```
It does this several times then all of Calamares freezes. I can exit Calamares, and I can tell both in Calamares and GParted the disk has been wiped and partitioned. Also, it looks like most of the OS was copied over. However, `/boot` is notably absent. It is unknown just how much else is missing.

This is being done in a VirtualBox VM with a development version of Drauger OS (currently based on Ubuntu 19.04. Will be rebased to 19.10 shortly after it's release). This is on a 64-bit system and the issues are being displayed both in EFI and BIOS. Disk is 50 GB and error is displayed no matter what the partition layout is set to.

**To Reproduce**
I am honestly unsure how to reproduce this. I could send the ISO to someone to have them test it or provide a download link. 

**Expected behavior**
I expect all files to be copied correctly

**Screenshots and Logs**
Terminal Output: https://pasteboard.co/IvvePTs.png 

**Additional context**
I have made sure rsync is installed, that the kernel supports selinux, and that I have the correct packages installed. I'm using Calamares 3.2.4-0ubuntu3. The OS is packaged in a squashfs as well if that is important.

**Comment by kkofler** [comment]
These errors happen if the destination file system does not support extended attributes such as the SELinux attributes. (E.g., it is normal to get those for the EFI System Partition. In your case, it is the target system's `/` though, which is more surprising.) But these errors will **not** prevent any part of the system from being copied, `rsync` skips only the extended attributes.

`/boot` not being copied is an entirely separate issue, and it is due to a bad `unpackfs.conf` configuration. By design, `unpackfs` copies only the data from one file system and not from other file systems mounted as child mount points. (It passes the `-x` = `--one-file-system` option to `rsync`.) You have to list those child mounts explicitly in `unpackfs.conf`.

So, in short, this is not a bug in Calamares, but a configuration issue. Please fix your `unpackfs.conf`.

**Comment by Batcastle** [comment]
While I appreciate that the lsetxattr error isn't an issue (Actually quite a relief. Lol), Defining `/boot` doesn't seem like an option since the entire OS is one massive squashfs, `/boot` and all. I even checked and `lsblk` reports in the live system that the entire thing is mounted as one. I will do some more investigation to figure out what the real problem is and make another issue if necessary. Thank you.

**Comment by kkofler** [comment]
We discussed this on IRC, and @adriaandegroot pointed out that you wrote that Calamares freezes, which is of course not normal and probably the real issue. (I had missed that important detail.) I guess files are missing because `rsync` is hanging up before they are copied.

If `/boot` is in the same file system as `/`, defining it separately is indeed not necessary.

I still think that this is a configuration or distro issue though, in that the freezes are probably due to rsyncing from a live system. The copy is probably hanging on some special file, either a kernel device node (e.g., if you're trying to copy `/dev/zero` or `/dev/urandom` or `/dev/random`) or a FUSE device node (e.g., gio-fuse is known to cause hangs if you try to copy its directory tree completely).

The kernel devices should really be in a separate `/dev` `tmpfs` (to be filled by `udev`), the FUSE devices should ideally not be mounted at all. But I strongly recommend just not attempting to copy the live system at all (I only know one distro doing that and its developers are aware of the pitfalls), but to unpack your base squashfs without the overlay (as most distros do) instead.

**Comment by Batcastle** [comment]
It might be due to miss-configuration, and that is indeed what I am hoping for. Right now, my `unpackfs.conf` looks like this:

```
---
unpack:
    -   source: "/run/live/medium/live/filesystem.squashfs"
        sourcefs: "squashfs"
        destination: ""
```
The path given is the the actual base squashfs file itself. Sorry for the miscommunication on that. The kernel device nodes (`/dev/null`, `/dev/stdin`, `/dev/random`, etc) all got copied before the hang. Although `/dev/stderr` doesn't look right.

**Comment by kkofler** [comment]
OK, so that brings us closer to finding the root cause. Files created or mounted on the running live image can't be it, because they won't be part of the squashfs image that is being unpackaged. Hence, the problem seems to be the content of the squashfs, or the target file system.

What are the contents of your `partition.conf` and `mount.conf`?

**Comment by Batcastle** [comment]
`mount.conf`:
```
extraMounts:
    - device: proc
      fs: proc
      mountPoint: /proc
    - device: sys
      fs: sysfs
      mountPoint: /sys
    - device: /dev
      mountPoint: /dev
      options: bind
    - device: tmpfs
      fs: tmpfs
      mountPoint: /run
    - device: /run/udev
      mountPoint: /run/udev
      options: bind

extraMountsEfi:
    - device: efivarfs
      fs: efivarfs
      mountPoint: /sys/firmware/efi/efivars
```
`partition.conf`:
```
efiSystemPartition: "/boot/efi"
enableLuksAutomatedPartitioning: false
neverCreateSwap: false
drawNestedPartitions: false
```

**Comment by kkofler** [comment]
And you are partitioning using the default file system, which should be `ext4` because your config file does not set anything else? Because if so, then your target `/` should definitely support extended attributes, so the errors are actually symptoms of a deeper issue. Are you sure partitioning is working?

**Comment by Batcastle** [comment]
Yes. I am sure partitioning is working and I am using the default partitioning (in testing I have just been clicking "Erase Drive" so I know it's completely reformatting the virtual drive). I can reboot the VM and mount the virtual drive manually and mess around with the files, this remains true even if there were no files before installation.

**Comment by Batcastle** [comment]
If this is an issue to do with `rsync`, would implementing what is discussed in #1118 fix this problem? Or, could this be an issue with the system's `selinux` implementation?

**Comment by kkofler** [comment]
Using `unsquashfs` instead of `rsync` may or may not fix this as a side effect. It might introduce other issues though.

This could be an SELinux issue indeed. If Calamares runs under a too restrictive SELinux context, that might explain the failures. (Do you see any AVC messages logged?) Calamares should really be running under a context like `sys_adm_t` that is allowed to do almost everything.

**Comment by Batcastle** [comment]
Okay. Well I am making sure that it has the correct SELinux context right now. For some reason it looks like it didn't have one set:
```
root@batcastle-Nitro-AN515-53:/# ls -Z $(which calamares)
? /usr/bin/calamares
```

Currently trying to figure out how to set it correctly. But that might at least partially explain the issues.

**Comment by Batcastle** [comment]
Didn't see any of those AVC messages. The thing is, selinux isn't even installed. I only have `libselinux1` installed. That being said, AppArmor is installed, so I don't know if that is related or not. But it is there if that matters

**Comment by Batcastle** [comment]
So I rebased to the Ubuntu 19.10 beta, so I am now on Calamares 3.2.12. I am testing in a VM right now. So far, not much different. Same `lsetxattr()` error, still freezing up when copying the contents of the squashfs file to the designated partitions, albeit it actually will close when I tell it to and not take several seconds in order to do so.

Is it possible I am not making my squashfs right and that's what is causing the issue? If this is possible (which I don't imagine it would be, just want to be sure) I can share the command used to make it.

**Comment by adriaandegroot** [comment]
I'm closing this with "feedback timeout", somewhat reluctantly. The current how-to-report-an-issue guide (from demmm) would have pushed harder for "give us an ISO link to check". There's a avoid-rsync approach via #1118 now, too.
